src/useAsnyc.js

```diff
@@ -0,0 +1,47 @@
+/* eslint react/prop-types: 0 */
+// @flow
+import * as React from 'react';
+import usePromise from './utils/usePromise';
+
+class AsyncStore {
+  constructor(result, error, state, lastValue) {
+    this._result = result;
+    this._error = error;
+    this._state = state;
+    this.lastValue = lastValue;
+  }
+
+  value() {
+    return this._result || null;
+  }
+
+  error() {
+    return this._error || null;
+  }
+
+  success() {
+    return this._state === 'resolved' ? this.value() : null;
+  }
+
+  failure() {
+    return this._state === 'rejected' ? this.error() : null;
+  }
+
+  initial() {
+    return this._state === 'initial';
+  }
+
+  loading() {
+    return this._state === 'pending';
+  }
+}
```

Comment:
I think we should simplify the interface and implementation (strangely by adding extra methods)

```diff
+ hasValue() -> boolean (depends on _latestValue)
+ isFailure() -> boolean (depends on _state)
+ isSuccess() -> boolean (depends on _state)
+ isLoading() -> boolean (depends on _state)
+ isInitial() -> boolean (depends on _state)
- initial
- loading
```

Then e.g. `success()` becomes `return this.isSuccess() ? this._result : undefined`